mapping_parenting_tech/analysis/prototyping/clustering_descriptions.py

```diff
@@ -68,34 +76,371 @@
+
+current_unassigned = len(s_df[s_df.s_cluster == -1].s_cluster.to_list())
+
+s_cluster_list = s_df.s_cluster.to_list()
+
+unassigned_indices = []
+unassigned_indices.extend([i for i, x in enumerate(s_cluster_list) if x == -1])
+
+X_list = [
+    x for i, x in enumerate(s_embedding_clustering) if i not in unassigned_indices
+]
+y_list = [i for i in s_cluster_list if i != -1]
+
+neigh.fit(X_list, y_list)
+
+for x in unassigned_indices:
+    s_df.iat[x, s_cluster_col] = neigh.predict([s_embedding_clustering[x]])
+
+print(f"{len(s_df[s_df.s_cluster == -1].s_cluster.to_list())} unassigned apps remain.")
+
+# %%
+# Visualise using altair (NB: -1=points haven't been assigned to a cluster)
+s_fig = (
+    alt.Chart(s_df.reset_index(), width=750, height=750)
+    .mark_circle(size=60)
+    .encode(
+        x="x", y="y", tooltip=["s_cluster", "appId", "summary"], color="s_cluster:N"
+    )
+).interactive()
+
+s_fig
+
+# %%
+s_relevant_clusters = set()
+for a_id in set(extra_apps + apps_of_interest):
+    if a_id in s_df.appId.values:
+        s_relevant_clusters.update(s_df[s_df.appId == a_id].s_cluster.values)
+        if s_df[s_df.appId == a_id].s_cluster.values == -1:
+            print(f"{a_id} not in cluster")
+
+s_relevant_clusters
+
+# %%
+s_cluster_options = {
+    "relevant": {
+        2: "stories/lullabies/music",
+        4: "learning through play / learning to code/ professional development",
+        6: "health / medical (adult)",
+        8: "animal-related (identify, sounds etc.)",
+        9: "play / make-belief",
+        10: "sleep (helping babies to)",
+        11: "education / parenting support",
+        17: "pregnancy",
+        18: "baby monitoring / development tracking (photos)",
+        19: "number games",
+        24: "learning games for toddlers",
+        26: "learning games (mostly ABCs)",
+        27: "learning ABCs",
+        28: "learning English (inc. as a foreign language)",
+    },
+    "interesting": {
+        1: "drawing / colouring",
+        12: "period / fertility tracking",
+        25: "learning games",
+    },
+    "unlikely": {
+        6: "student learning",
+        16: "meditation / calm / sleep (for adults)",
+        23: "mental health",
+    },
+    "discard": {
+        0: "Setting light levels (eg screen brightness)",
+        3: "driving test",
+        5: "engineering",
+        7: "connecting with people",
+        13: "personal organisation",
+        14: "child safety / web filtering",
+        15: "outdoor / mapping",
+        20: "religion",
+        21: "minfulness / Buddhism",
+        22: "motivation",
+    },
+}
+
+# %%
+cluster_relevance_ = list()
+s_cluster_relevance_ = list()
+
+cluster_type_ = list()
+s_cluster_type_ = list()
+
+for i, app_id in enumerate(df.appId.tolist()):
+    relevance, purpose = get_cluster_info(int(df.cluster.iloc[i]), cluster_options)
+    cluster_relevance_.append(relevance)
+    cluster_type_.append(purpose)
+
+    relevance, purpose = get_cluster_info(
```

Comment:
This line threw a TypeError
```
TypeError: cannot unpack non-iterable int object
```

Reply:
@beingkk Can you try again with the latest commit and data from `outputs` and `inputs`? Think this should work now?

Reply:
Hey, I still got the error - but I've proposed a fix in the function definition. The issue that you're expecting here two output variables, but there are cases when your function returns only a single output (ie, `return 0`)